nifi-external/nifi-kafka-connect/nifi-kafka-connector/src/main/java/org/apache/nifi/kafka/connect/StatelessNiFiSourceTask.java

```diff
@@ -0,0 +1,275 @@
+            final String portName = entry.getKey();
+            final List<FlowFile> flowFiles = entry.getValue();
+
+            if (!flowFiles.isEmpty() && !expectedPortName.equals(portName)) {
+                logger.error("Dataflow transferred FlowFiles to Port {} but was expecting data to be transferred to {}. Rolling back session.", portName, expectedPortName);
+                trigger.cancel();
+                throw new RetriableException("Data was transferred to unexpected port");
+            }
+        }
+    }
+
+
+    private SourceRecord createSourceRecord(final FlowFile flowFile, final byte[] contents) {
+        final Map<String, ?> partition = Collections.emptyMap();
+        final Map<String, ?> sourceOffset = Collections.emptyMap();
+        final Schema valueSchema = (contents == null || contents.length == 0) ? null : Schema.BYTES_SCHEMA;
+
+        // Kafka Connect currently gives us no way to determine the number of partitions that a given topic has.
+        // Therefore, we have no way to partition based on an attribute or anything like that, unless we left it up to
+        // the dataflow developer to know how many partitions exist a priori and explicitly set an attribute in the range of 0..max,
+        // but that is not a great solution. Kafka does support using a Simple Message Transform to change the partition of a given
+        // record, so that may be the best solution.
+        final Integer topicPartition = null;
+
+        final String topic;
+        if (topicNameAttribute == null) {
+            topic = topicName;
+        } else {
+            final String attributeValue = flowFile.getAttribute(topicNameAttribute);
+            topic = attributeValue == null ? topicName : attributeValue;
+        }
+
+        final List<Header> headers;
+        if (headerAttributeNamePattern == null) {
+            headers = Collections.emptyList();
+        } else {
+            headers = new ArrayList<>();
+
+            for (final Map.Entry<String, String> entry : flowFile.getAttributes().entrySet()) {
+                if (headerAttributeNamePattern.matcher(entry.getKey()).matches()) {
+                    final String headerName = entry.getKey();
+                    final String headerValue = entry.getValue();
+
+                    headers.add(new StatelessNiFiKafkaHeader(headerName, headerValue));
+                }
+            }
+        }
+
+        final Object key = keyAttributeName == null ? null : flowFile.getAttribute(keyAttributeName);
+        final Schema keySchema = key == null ? null : Schema.STRING_SCHEMA;
+        final Long timestamp = System.currentTimeMillis();
+
+        return new SourceRecord(partition, sourceOffset, topic, topicPartition, keySchema, key, valueSchema, contents, timestamp, headers);
+    }
+
+    @Override
+    public void commitRecord(final SourceRecord record, final RecordMetadata metadata) throws InterruptedException {
```

Comment:
This method is usually called by a producer network thread when the record batch was ACK'd by the broker.
It is possible that poll is called multiple times before the commitRecord is invoked with the records returned by the first call of poll. Because of this, the result cached in triggerResult might be overwritten, and never get acknowledged.
The task might need to track multiple TriggerResults to be able to meaningfully acknowledge them.

Reply:
Stateless framework is going to prevent a second invocation until the previous one completes. But I can easily determine if the previous invocation is still working and if so, return `null` from poll().